fix: In-scene NetworkObjects disabled but never re-enabled after scene transition

NoelStephensUnity opened the pull request:
Only when there is an in-scene placed NetworkObject that moves itself into the DDOL scene and there is a full scene transition (i.e. LoadSceneMode.Single) the in-scene placed NetworkObject will get disabled and never re-enabled.
This PR resolves the above issue.

[MTT-1479](https://jira.unity3d.com/browse/MTT-1479)

### PR Checklist
- [X] Have you added a backport label (if needed)? For example, the `type:backport-release-*` label. After you backport the PR, the label changes to `stat:backported-release-*`.
- [X] Have you updated the changelog? Each package has a `CHANGELOG.md` file.

## Changelog
### com.unity.netcode.gameobjects
- Fixed in-scene NetworkObjects that are moved into the DDOL scene not getting restored to their original active state (enabled/disabled) after a full scene transition (i.e. LoadSceneMode.Single).
- Fixed all NetworkObjects marked to persist a full scene transition now are restored to their original active state (enabled/disabled) after a full scene transition (i.e. LoadSceneMode.Single).

## Testing and Documentation
* Includes unit tests.
* No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

This is the fix for the in-scene NetworkObjects moved into DDOL and the after a scene transition they become disabled but never re-enabled.

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
test

Had to make some private methods internal for testing purposes.
Adding test to validate the fix.

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
style

white space fixes.

NoelStephensUnity committed:
style

Adding comment above the primary line of code that is the fix for this PR

NoelStephensUnity committed:
style

removing unused namespace.

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
style

removing several LFs

JesseOlmer commented on testproject/Assets/Tests/Runtime/NetworkSceneManagerTests.cs:
```diff
@@ -935,7 +936,143 @@ public IEnumerator FastReaderAllocationTest()
+        private TestingState m_TestingState;
+
+        [UnityTest]
+        public IEnumerator DDOLMoveAndRestoreInSceneNetworkObject()
```
I can't tell what this test does by its name.
I also can't really tell what the test does by looking at the test method. Is it testing that `MoveObjectsToDontDestroyOnLoad` sets them to inactive? Is it testing that `MoveObjectsFromDontDestroyOnLoadToScene` sets them to active? Is it testing both?

NoelStephensUnity replied:
Yeah... I need to come up with a better name for that test probably.
The test uses those two methods when doing a full scene transition (i.e. LoadSceneMode.Single).
- Pre-Scene Transition--> MoveObjectsToDontDestroyOnLoad (disables in-scene placed NOs in DDOL)
- Load Scene
- Post-Scene Transition--> MoveObjectsFromDontDestroyOnLoadToScene (enables in-scene placed NOs in DDOL)

I could just have easily manually disabled the NetworkObject and then just called the MoveObjectsFromDontDestroyOnLoadToScene method to validate the fix, but figured I might as well include the initial method that disables the in-scene placed NetworkObject (MoveObjectsToDontDestroyOnLoad) since that is what disables the NetworkObject(s) prior to doing the full scene transition.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1830,7 +1830,7 @@ internal void HandleSceneEvent(ulong clientId, FastBufferReader reader)
         /// the "Do not destroy on load" scene.
         /// </summary>
-        private void MoveObjectsToDontDestroyOnLoad()
+        internal void MoveObjectsToDontDestroyOnLoad()
```
It appears that these are being made internal so that you can test them directly? This is a testing anti-pattern. To paraphrase a great number of testing authorities: "How do you test internals? Through the public API."

NoelStephensUnity replied:
Yeah... the test to validate this fix was a little tricky.
The only place the two methods pertaining to this bug and fix (MoveObjectsToDontDestroyOnLoad and MoveObjectsFromDontDestroyOnLoadToScene ) are called is when there is a full scene transition taking place.  Since we cannot do a full scene transition (i.e. it would unload the Test Runner's temporary scene and would break the rest of the tests) in a unit and/or integration test, the only way to write a test (currently) to verify the fix in this PR was to create the same circumstances and call the same methods...which required them to be accessible.
If you want, I could reduce the anti-pattern footprint by just testing the MoveObjectsFromDontDestroyOnLoadToScene method and not including the MoveObjectsToDontDestroyOnLoad in that process (i.e. just disable it in the unit test itself).

JesseOlmer commented on testproject/Assets/Tests/Runtime/NetworkSceneManagerTests.cs:
```diff
@@ -935,7 +936,143 @@ public IEnumerator FastReaderAllocationTest()
+
+            private void Start()
+            {
+                DontDestroyOnLoad(gameObject);
```
nit: since DontDestroyOnLoad is part of your test case it increases readability to put this in your test setup, not down here in a utility class

NoelStephensUnity replied:
Well, actually the test case was based on one of our manual tests that also included the NotifyClientRpc component:
```
    public class NotifyClientRpc : NetworkBehaviour
    {
        private static NotifyClientRpc s_Singleton;

        private void OnEnable()
        {
            if (s_Singleton != null && s_Singleton != this)
            {
                DestroyImmediate(gameObject);
                return;
            }
            DontDestroyOnLoad(this);
        }
...
```
This component was how I initially found the bug when running through the manual tests, and since it moved itself into the DDOL when enabled I went ahead and tried to replicate the same scenario:
> Only when there is an in-scene placed NetworkObject that moves itself into the DDOL scene and there is a full scene transition

Does this make more sense why I am doing that in the component and not in the test setup?

NoelStephensUnity replied:
Made changes to this so resolving this conversation as it no longer applies

JesseOlmer commented on testproject/Assets/Tests/Runtime/NetworkSceneManagerTests.cs:
```diff
@@ -935,7 +936,143 @@ public IEnumerator FastReaderAllocationTest()
+
+        public class ForcedInSceneBehaviour : NetworkBehaviour
+        {
+            public event Action OnNetworkObjectSpawned;
```
nit: this event system to generic condition tracker feels like overkill. Why not just have an `InSceneNetworkObjectDDOLTest` field on the network behaviour, assign it when you create it, and have it do something like `InSceneNetworkObjectDDOLTest.OnEnabledWasCalled = true;`

Then, in the test case, instead of the "waitforcondition" and "conditionmet(false)" stuff you can just do the waiting in-place and increase readability of the test:

```
            while (!OnEnabledWasCalled)
            {
                yield return new WaitForSeconds(m_ConditionMetFrequency);
            }
```

NoelStephensUnity replied:
Sure, I could just do something like:
```
            while (networkObject.isActiveAndEnabled)
            {
                yield return new WaitForSeconds(m_ConditionMetFrequency);
            }
```
and to check it is re-enabled
```
            while (!networkObject.isActiveAndEnabled)
            {
                yield return new WaitForSeconds(m_ConditionMetFrequency);
            }
```
in place of the other stuff.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1916,6 +1916,8 @@ private void MoveObjectsFromDontDestroyOnLoadToScene(Scene scene)
                 if (sobj.gameObject.scene == DontDestroyOnLoadScene && (sobj.IsSceneObject == null || sobj.IsSceneObject.Value))
                 {
+                    // set in-scene placed NetworkObjects back to active
+                    sobj.gameObject.SetActive(true);
```
Is this going to enabled objects which the user disabled manually?

JesseOlmer replied:
Maybe worth adding a few test cases:
- Scene-placed objects not in DDOL shouldn't be impacted
- Dynamic objects not in DDOL shouldn't be impacted
- Scene-placed objects in DDOL which are active should stay active
- Scene-Placed objects in DDOL which are inactive should stay inactive
- Dynamic objects in DDOL which are active shouldn't be impacted

NoelStephensUnity replied:
I can do that.

NoelStephensUnity replied:
The most recent updated test includes the following test cases where all NetworkObjects' DestroyWithScene are set to false:

- Tests for both default enabled and disabled cases:
  - Scene-placed NetworkObjects not already in the DDOL
  - Dynamically spawned NetworkObjects not already in the DDOL
  - Scene-placed NetworkObjects already in the DDOL
  - Dynamically spawned NetworkObjects already in the DDOL

_As a side note:
Both dynamically spawned and in-scene placed NetworkObjects that do not have the DestroyWithScene set to false will be despawned before the scene transition._

andrews-unity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
fix and test

This now tracks the state of NetworkObjects when a full scene transition takes place.

This includes a refactored test for NetworkObjects (dynamically spawned and in-scene placed) that persist a full scene transition.

NoelStephensUnity committed:
test update

This makes sure that DestroyWithScene is set to false for all InSceneNetworkObjectState tests.

NoelStephensUnity committed:
style

Removing whitespace.

NoelStephensUnity committed:
updated changelog

Added entries for the fixes applied in this PR.

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -367,6 +367,11 @@ public class NetworkSceneManager : IDisposable
 
+        // For full scene transitions we want to keep track of in-scene placed NetworkObjects' active state that are moved into the DDOL
+        // by user code.
+        private Dictionary<ulong, bool> m_NetworkObjectDDOLActiveState = new Dictionary<ulong, bool>();
```
Can you add a comment stating what the key is ? (I guess the NetworkObjectId)

NoelStephensUnity replied:
Removing the state tracking and leaving it up to the user as to whether they want to disable during full scene transition or not.

MFatihMAR commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -367,6 +367,11 @@ public class NetworkSceneManager : IDisposable
 
+        // For full scene transitions we want to keep track of in-scene placed NetworkObjects' active state that are moved into the DDOL
+        // by user code.
+        private Dictionary<ulong, bool> m_NetworkObjectDDOLActiveState = new Dictionary<ulong, bool>();
+        private Dictionary<ulong, bool> m_NetworkObjectUserMovedDDOLActiveState = new Dictionary<ulong, bool>();
```
why do we need 2 dictionaries? can we have just 1?

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -1830,21 +1835,40 @@ internal void HandleSceneEvent(ulong clientId, FastBufferReader reader)
+                // Determine if we are already in the DDOL (i.e. moved by user)
+                bool isUserMovedNetworkObjectInDDOL = sobj.gameObject.scene == DontDestroyOnLoadScene;
+                if (!sobj.DestroyWithScene || isUserMovedNetworkObjectInDDOL)
                 {
                     // Only move objects with no parent as child objects will follow
                     if (sobj.gameObject.transform.parent == null)
                     {
-                        UnityEngine.Object.DontDestroyOnLoad(sobj.gameObject);
-                        // Since we are doing a scene transition, disable the GameObject until the next scene is loaded
-                        sobj.gameObject.SetActive(false);
+                        // If this is network object was moved into the DDOL by the user
+                        if (isUserMovedNetworkObjectInDDOL)
+                        {
+                            // Keep track of its active state in the m_NetworkObjectUserMovedDDOLActiveState list
+                            m_NetworkObjectUserMovedDDOLActiveState.Add(sobj.NetworkObjectId, sobj.gameObject.activeInHierarchy);
+
+                            // Since we are doing a scene transition, disable the GameObject until the next scene is loaded (LoadSceneMode.Single only)
+                            sobj.gameObject.SetActive(false);
+                        }
```
just to note here: based on our internal conversation, we want to discover what would happen if we never touch object's active state at all — in other terms, what if we go back and delete `SetActive(false)` code we had there before.
if leaving objects as they are during scene transition wouldn't cause any issues internal to the SDK, we could simply not touch it and leave it up to user to decide how they want to handle objects ticking during scene transition.

NoelStephensUnity committed:
refactor

After peer review during PR Triage we all decided to not change or track a NetworkObject's parent GameObject's state and to not disable during full scene transition.

Removed all active state tracking dictionaries and references as well as removed disabling DDOL migrated GameObjects from the runtime NGO code base.

Adjusted the DDOL test to check the state is maintained throughout the scene transition.

The DDOL test only required a minor

NoelStephensUnity committed:
updated changelog

Updating the changelog to reflect most recent updates.

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

0xFA11 reviewed (approved):
🚀

NoelStephensUnity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

github-actions[bot] committed:
Merge develop into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1354) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1354) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=1354) it.</sub>

andrews-unity committed:
Merge branch 'develop' into fix/InScene-NetworkObjects-DDOL-Disabled-MTT-1479
